diff -ruN a/vendor/github.com/tsg/gopacket/pcap/pcap.go b/vendor/github.com/tsg/gopacket/pcap/pcap.go
--- a/vendor/github.com/tsg/gopacket/pcap/pcap.go	2017-01-28 02:37:00.882796778 +0000
+++ b/vendor/github.com/tsg/gopacket/pcap/pcap.go	2017-02-05 03:09:03.001182695 +0000
@@ -8,7 +8,15 @@
 package pcap
 
 /*
-#cgo linux LDFLAGS: -lpcap
+#cgo linux,386 LDFLAGS: /libpcap/i386/usr/lib/i386-linux-gnu/libpcap.a
+#cgo linux,amd64 LDFLAGS: /libpcap/amd64/usr/lib/x86_64-linux-gnu/libpcap.a
+#cgo linux,arm LDFLAGS: /libpcap/arm/usr/lib/arm-linux-gnueabihf/libpcap.a
+#cgo linux,arm64 LDFLAGS: /libpcap/arm64/usr/lib/aarch64-linux-gnu/libpcap.a
+#cgo linux,mips LDFLAGS: /libpcap/mips/usr/lib/mips-linux-gnu/libpcap.a
+#cgo linux,mips64 LDFLAGS: /libpcap/mips64/usr/lib/mips64-linux-gnuabi64/libpcap.a
+#cgo linux,mipsle LDFLAGS: /libpcap/mipsle/usr/lib/mipsel-linux-gnu/libpcap.a
+#cgo linux,mips64le LDFLAGS: /libpcap/mips64le/usr/lib/mips64el-linux-gnuabi64/libpcap.a
+#cgo linux,ppc64le LDFLAGS: /libpcap/ppc64le/usr/lib/powerpc64le-linux-gnu/libpcap.a
 #cgo freebsd LDFLAGS: -lpcap
 #cgo openbsd LDFLAGS: -lpcap
 #cgo darwin LDFLAGS: -lpcap
